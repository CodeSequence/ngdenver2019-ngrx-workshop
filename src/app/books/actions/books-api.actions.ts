import { Book } from "src/app/shared/models/book.model";
import { Action } from "@ngrx/store";

export enum BooksApiActionTypes {
  BooksLoaded = '[Books API] Books Loaded Success',
}

export class BooksLoaded implements Action {
  readonly type = BooksApiActionTypes.BooksLoaded;

  constructor(public books: Book[]) {}
}

export type BooksApiActions = 
  | BooksLoaded;
