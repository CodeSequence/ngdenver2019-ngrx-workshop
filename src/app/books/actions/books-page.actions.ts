import { Book } from "src/app/shared/models/book.model";
import { Action } from "@ngrx/store";

export enum BooksActionTypes {
  SelectBook = "[Books Page] Select Book",
  ClearSelectedBook = "[Books Page] Clear Selected Book",
  CreateBook = "[Books Page] Create Book",
  UpdateBook = "[Books Page] Update Book",
  DeleteBook = "[Books Page] Delete Book"
}

export class SelectBook implements Action {
  readonly type = BooksActionTypes.SelectBook;

  constructor(public bookId: string) {}
}

export class ClearSelectedBook implements Action {
  readonly type = BooksActionTypes.ClearSelectedBook;
}

export class CreateBook implements Action {
  readonly type = BooksActionTypes.CreateBook;

  constructor(public book: Book) {}
}

export class UpdateBook implements Action {
  readonly type = BooksActionTypes.UpdateBook;

  constructor(public book: Book) {}
}

export class DeleteBook implements Action {
  readonly type = BooksActionTypes.DeleteBook;

  constructor(public book: Book) {}
}

export type BooksActions =
  | SelectBook
  | ClearSelectedBook
  | CreateBook
  | UpdateBook
  | DeleteBook;
