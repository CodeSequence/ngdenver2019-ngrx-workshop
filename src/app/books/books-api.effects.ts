import { Injectable } from "@angular/core";
import { Effect, Actions, ofType } from "@ngrx/effects";
import { BooksPageActions, BooksApiActions } from "./actions";
import { BooksService } from "../shared/services/book.service";
import { mergeMap, map, catchError } from "rxjs/operators";
import { EMPTY } from "rxjs";

@Injectable()
export class BooksApiEffects {
  @Effect()
  loadBooks$ = this.actions$.pipe(
    ofType(BooksPageActions.BooksActionTypes.Enter),
    mergeMap(() =>
      this.booksService.all().pipe(
        map(books => new BooksApiActions.BooksLoaded(books)),
        catchError(() => EMPTY)
      )
    )
  );

  constructor(
    private booksService: BooksService,
    private actions$: Actions<
      BooksPageActions.BooksActions | BooksApiActions.BooksApiActions
    >
  ) {}
}
