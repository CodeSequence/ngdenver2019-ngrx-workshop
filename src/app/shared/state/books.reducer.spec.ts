import { BooksApiActions, BooksPageActions } from "src/app/books/actions";
import { Book } from "../models/book.model";
import {
  reducer,
  initialState,
  selectActiveBook,
  adapter,
  selectAll
} from "./books.reducer";

describe("Books Reducer", () => {
  it("should return the initial state when initialized", () => {
    const state = reducer(undefined, { type: "@@init" } as any);

    expect(state).toBe(initialState);
  });

  it("should load all books when the API loads them all successfully", () => {
    const books: Book[] = [{ id: "1", name: "Castaway", earnings: 1000000 }];
    const action = BooksApiActions.booksLoaded({ books });

    const state = reducer(initialState, action);

    expect(state).toMatchSnapshot();
  });

  it("should add a newly created book to the state", () => {
    const book: Book = { id: "1", name: "Forrest Gump", earnings: 200000000 };
    const action = BooksApiActions.bookCreated({ book });

    const state = reducer(initialState, action);

    expect(state).toMatchSnapshot();
  });

  it("should remove books from the state when they are deleted", () => {
    const book: Book = { id: "1", name: "Apollo 13", earnings: 1000 };
    const firstAction = BooksApiActions.bookCreated({ book });
    const secondAction = BooksApiActions.bookDeleted({ book });

    const state = [firstAction, secondAction].reduce(reducer, initialState);

    expect(state).toMatchSnapshot();
  });

  describe("Selectors", () => {
    const initialState = { activeBookId: null, ids: [], entities: {} };

    describe("selectActiveBook", () => {
      it("should return null if there is no active book", () => {
        const result = selectActiveBook(initialState);

        expect(result).toBe(null);
      });

      it("should return the active book if there is one", () => {
        const book: Book = { id: "1", name: "Castaway", earnings: 1000000 };
        const state = adapter.addAll([book], {
          ...initialState,
          activeBookId: "1"
        });
        const result = selectActiveBook(state);

        expect(result).toBe(book);
      });
    });

    describe("selectAll", () => {
      it("should return all the loaded books", () => {
        const books: Book[] = [
          { id: "1", name: "Castaway", earnings: 1000000 }
        ];
        const state = adapter.addAll(books, initialState);
        const result = selectAll(state);

        expect(result.length).toBe(1);
      });
    });
  });
});
