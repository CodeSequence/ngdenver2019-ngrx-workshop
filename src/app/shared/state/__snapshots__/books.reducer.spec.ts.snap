// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Books Reducer should add a newly created book to the state 1`] = `
Object {
  "activeBookId": "1",
  "entities": Object {
    "1": Object {
      "earnings": 200000000,
      "id": "1",
      "name": "Forrest Gump",
    },
  },
  "ids": Array [
    "1",
  ],
}
`;

exports[`Books Reducer should load all books when the API loads them all successfully 1`] = `
Object {
  "activeBookId": null,
  "entities": Object {
    "1": Object {
      "earnings": 1000000,
      "id": "1",
      "name": "Castaway",
    },
  },
  "ids": Array [
    "1",
  ],
}
`;

exports[`Books Reducer should remove books from the state when they are deleted 1`] = `
Object {
  "activeBookId": null,
  "entities": Object {},
  "ids": Array [],
}
`;
